src/transformStateWithClones.js

```diff
@@ -7,7 +7,48 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let stateCopy = { ...state };
+
+  const addProperties = (extraData) => {
+    stateCopy = Object.assign(stateCopy, extraData);
+    stateHistory.push({ ...stateCopy });
```

Comment:
Pushing the state copy to 'stateHistory' is repeated in every action handler. Consider doing this once after the action handling to adhere to the 'Don't Repeat Yourself' (DRY) principle.